src/Lib/App/Error.hs

```diff
@@ -1,10 +1,104 @@
+{-# LANGUAGE ConstraintKinds #-}
+
 module Lib.App.Error
        ( AppError (..)
+       , WithError
+       , IError
+       , throwOnNothingM
+       , notFoundOnNothingM
+       , isServerError
+       , isNotAllowed
+       , isInvalid
+       , notFound
+       , serverError
+       , notAllowed
+       , invalid
+       , headerDecodeError
+       , jobDecodeError
+       , toHttpError
        ) where
 
-data AppError =
-    Invalid Text
-  | NotAllowed Text
-  | NotFound
-  | ServerError Text
-  deriving (Show, Eq)
+import Control.Monad.Except (MonadError, throwError)
+import Servant.Server (ServantErr, err401, err404, err417, err500, errBody)
+
+newtype AppError = InternalError IError deriving (Show, Eq)
+
+data IError =
+    -- | General not found
+      NotFound
+    -- | Some exceptional circumstance has happened
+    -- stop execution and return. Optional text to
+    -- provide some context in server logs
+    | ServerError Text
+    -- | A required permission level was not met.
+    -- Optional text to provide some context.
+    | NotAllowed Text
+    -- | Given inputs do not conform to the expected
+    -- format or shape. Optional text to
+    -- provide some context in server logs
+    | Invalid Text
+    -- | An authentication header that was required
+    -- was provided but not in a format that the server
+    -- can understand
+    | HeaderDecodeError
+    -- | Failed to parse 'CDMP.Effects.Job.Job' from 'Text'.
+    | JobDecodeError Text
```

Comment:
Documentation for this constructor says that this is something `CDMP` specific. So shouldn't exist here.